src/domain/mongoose_domain_core.erl

```diff
@@ -0,0 +1,139 @@
+%% Generally, you should not call anything from this module.
+%% Use mongoose_domain_api module instead.
+-module(mongoose_domain_core).
+-include("mongoose_logger.hrl").
+
+-export([start/1, stop/0]).
+-export([start_link/1]).
+-export([get_host_type/1]).
+-export([is_locked/1]).
+
+%% API for tests
+-export([dump/0, restore/1]).
+
+%% API, used by DB module
+-export([remove_all_unlocked/0,
+         upsert_unlocked/2,
+         remove_unlocked/1]).
+
+-export([get_all_locked/0,
+         get_domains_by_host_type/1]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(TABLE, ?MODULE).
+
+start(Pairs) ->
+    ChildSpec =
+        {?MODULE,
+         {?MODULE, start_link, [Pairs]},
+         permanent, infinity, worker, [?MODULE]},
+    just_ok(supervisor:start_child(ejabberd_sup, ChildSpec)).
+
+%% For tests
+stop() ->
+    supervisor:terminate_child(ejabberd_sup, ?MODULE),
+    supervisor:delete_child(ejabberd_sup, ?MODULE),
+    ok.
+
+start_link(Pairs) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Pairs], []).
+
+get_host_type(Domain) ->
+    case ets:lookup(?TABLE, Domain) of
+        [] ->
+            {error, not_found};
+        [{_Domain, HostType, _Locked}] ->
+            {ok, HostType}
+    end.
+
+is_locked(Domain) ->
+    case ets:lookup(?TABLE, Domain) of
+        [{_Domain, _HostType, _Locked = true}] ->
+            true;
+        _ ->
+            false
+    end.
+
+remove_all_unlocked() ->
+    ets:match_delete(?TABLE, {'_', '_', false}).
+
+get_all_locked() ->
+    heads(ets:match(?TABLE, {'$1', '_', true})).
+
+get_domains_by_host_type(HostType) when is_binary(HostType) ->
+    heads(ets:match(?TABLE, {'$1', HostType, '_'})).
+
+heads(List) ->
+    [H || [H|_] <- List].
+
+upsert_unlocked(Domain, HostType) ->
+    case is_locked(Domain) of
+        true ->
+            %% Ignore any locked domains
+            ?LOG_ERROR(#{what => domain_locked_but_in_db, domain => Domain});
+        false ->
+            ets:insert(?TABLE, new_object(Domain, HostType, false))
+    end,
+    ok.
+
+remove_unlocked(Domain) ->
```

Comment:
Remove should accept domain/host_type pair to avoid any race conditions.

Reply:
And what is the behaviour in the case of race condition detected?
DB does not know, what what the host_type of the removed domain.

Reply:
report error! do not remove domain!

Reply:
Again.
DB has information about the fact that a domain is removed. 
If a domain is removed from DB, DB has no information about the domain type. It's `null`.

Reply:
ok, look. we have domain/host_type pair as input parameter, the behaviour of the function should  be like this:

- there is such domain/host_type pair and we can remove it (it's not configured on init) - we remove it and return success
- there is such domain/host_type pair and we can not remove it (it is configured on init) - we do nothing and return a corresponding error.
- there is such domain but host_type is not matching - we do nothing and return a corresponding error. (MB different errors for domains configured dynamically and on init)
- there's no such domain - we do nothing and return success

Reply:
ok, it would require rework of DB layer/schema/tests.
Once domain is removed, we would not be able to provide hosttype from db.

Reply:
I'm not sure i understand how the core module is related to DB storage.

Reply:
After in person discussion I understand better the problem highlighted by @arcusfelis in the previous comment:
```
Event table doesn't store the kind of changes but only the fact that some domain has changed.
it's joined with the settings table (LEFT JOIN on domain field) and if settings table has no 
domain record anymore then returned host type field is "null". 
 ```
 
of course, we can check the host type from the ETS table, but it doesn't really make any sense to get the host type to just remove it.  So as agreed, we will have domain/host_type pair input parameter for `mongoose_domain_api` interface only, `mongoose_domain_core` will use only domain name for removal. anyway `mongoose_domain_core` module must be used only by the domains service modules.

@arcusfelis thanks for the good and very detailed explanation.